<a href="https://colab.research.google.com/github/Vitela-debug/Colab_S1_Clase-de-progralaci-n_273522_TVB/blob/main/Sesion12_Evaluacion_Analisis_Categoricos_TVB_273522.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Programación para Analítica Descriptiva y Predictiva**
**Maestría en Inteligencia Artificial y Analítica de Datos**

# Sesión 12: Evaluación — Limpieza y Transformación de Datos Categóricos

**Entrega individual**

- **Nombre**: Tania Vitela Barraza
- **Matrícula**: 273522

Esta evaluación aplica los tres temas de la Sesión 12 (errores tipográficos y valores inconsistentes, alta cardinalidad, tipos incorrectos) a un dataset que no se trabajó en clase: **Telco Customer Churn**.


No hay una única respuesta correcta en varias de las actividades — lo que se evalúa es que la conclusión esté respaldada por el código que la sustenta, no solo la conclusión en sí.

## Preparación

In [ ]:
import pandas as pd
import numpy as np

url = 'https://raw.githubusercontent.com/treselle-systems/customer_churn_analysis/master/WA_Fn-UseC_-Telco-Customer-Churn.csv'
df = pd.read_csv(url)
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


---
## Actividad 1 — Formato y valores inconsistentes (15 pts)

Revisa **todas** las columnas categóricas del dataset (no elijas solo una) en busca de variantes de formato (mayúsculas, espacios) que deberían normalizarse.

In [ ]:
# 1.1 — Recorre todas las columnas categóricas con .value_counts() o .unique()
# para inspeccionar sus valores.

columnas_categoricas = df.select_dtypes(include='object').columns # Obtener las columnas categóricas

# Mostrar los valores de cada columna categórica mediante un for para inspeccionar todos los valores
for columna in columnas_categoricas:
    print(f"\n==={columna} ===")
    print(df[columna].value_counts())

#Este código revisa todas las columnas categóricas y muestra sus valores y
#cuántas veces aparecen, para detectar diferencias en la forma de escribirlos.



===customerID ===
customerID
3186-AJIEK    1
7590-VHVEG    1
5575-GNVDE    1
8775-CEBBJ    1
2823-LKABH    1
             ..
6713-OKOMC    1
1452-KIOVK    1
9305-CDSKC    1
9237-HQITU    1
7795-CFOCW    1
Name: count, Length: 7043, dtype: int64

===gender ===
gender
Male      3555
Female    3488
Name: count, dtype: int64

===Partner ===
Partner
No     3641
Yes    3402
Name: count, dtype: int64

===Dependents ===
Dependents
No     4933
Yes    2110
Name: count, dtype: int64

===PhoneService ===
PhoneService
Yes    6361
No      682
Name: count, dtype: int64

===MultipleLines ===
MultipleLines
No                  3390
Yes                 2971
No phone service     682
Name: count, dtype: int64

===InternetService ===
InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

===OnlineSecurity ===
OnlineSecurity
No                     3498
Yes                    2019
No internet service    1526
Name: count, dtype: int64

===OnlineBackup ===
Online

**1.2 — Conclusión (responde aquí en Markdown):**

¿Encontraste alguna columna con inconsistencias de formato? Si sí, ¿cuál y qué código usarías para corregirla? Si no encontraste ninguna, dilo explícitamente — es una conclusión válida siempre que esté respaldada por lo que revisaste en 1.1.

_Tu respuesta:_

No encontré inconsistencias de formato en las columnas categóricas. Al revisar los valores, se puede observar que están escritos de manera uniforme, sin diferencias de mayúsculas, minúsculas o espacios. Por ejemplo, los valores de gender aparecen como Male y Female, y en otras columnas se utilizan correctamente Yes y No. Por lo tanto, no es necesario aplicar ningún código para corregir el formato. En caso de requerir utilizaría herramientas como:`.str.strip().str.lower()`

---
## Actividad 2 — Valores inválidos (20 pts)

Varias columnas de este dataset (`OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, `StreamingMovies`) tienen un tercer valor además de `'Yes'`/`'No'`: `'No internet service'`. De forma similar, `MultipleLines` tiene `'No phone service'`.

In [ ]:
df['OnlineSecurity'].value_counts()

,count
OnlineSecurity,
No,3498
Yes,2019
No internet service,1526


In [ ]:
# 2.1 — Verifica: ¿las filas con 'No internet service' en OnlineSecurity
# coinciden con las filas donde InternetService == 'No'?
# (pista: cruza ambas columnas con pd.crosstab o filtrando)

#Se definen las columnas con un tercer valor
columnas_internet = ['OnlineSecurity','OnlineBackup','DeviceProtection','TechSupport','StreamingTV','StreamingMovies']

for columna in columnas_internet:
    print(f"\n=== {columna} ===")
    print(pd.crosstab(df[columna], df['InternetService']))

  #El código sirve para comprobar que los datos de los servicios de internet sean coherentes con el tipo de internet
  #que tiene cada cliente. Con for revisa cada servicio y con pd.crosstab() compara sus valores con InternetService



=== OnlineSecurity ===
InternetService       DSL  Fiber optic    No
OnlineSecurity                              
No                   1241         2257     0
No internet service     0            0  1526
Yes                  1180          839     0

=== OnlineBackup ===
InternetService       DSL  Fiber optic    No
OnlineBackup                                
No                   1335         1753     0
No internet service     0            0  1526
Yes                  1086         1343     0

=== DeviceProtection ===
InternetService       DSL  Fiber optic    No
DeviceProtection                            
No                   1356         1739     0
No internet service     0            0  1526
Yes                  1065         1357     0

=== TechSupport ===
InternetService       DSL  Fiber optic    No
TechSupport                                 
No                   1243         2230     0
No internet service     0            0  1526
Yes                  1178          866     0

=== St

**2.2 — Conclusión (responde aquí en Markdown):**

¿`'No internet service'` es un valor inválido (como `Absurd`/`YOLO` en la sesión de clase) o es una categoría legítima? Justifica tu respuesta con lo que verificaste en 2.1. ¿Tomarías alguna acción sobre esta columna, o la dejarías tal cual?

_Tu respuesta:_

No internet service es una categoría legítima, no un valor inválido. En 2.1 comprobamos que las 1526 filas con No internet service en todas las distintas columnas coinciden con InternetService = No. Por lo tanto, este valor tiene un significado y es consistente con los datos. Yo dejaría la columna tal como está y no eliminaría ni cambiaría ese valor.

---
## Actividad 3 — Alta cardinalidad (25 pts)

In [ ]:
# 3.1 — Calcula .nunique() para TODAS las columnas del dataset (no solo las categóricas)
# y la razón (valores únicos / total de filas) para cada una.

columnas = df.columns.tolist()

for col in columnas:
    n_unicos = df[col].nunique()
    ratio = n_unicos / len(df)
    print(f'{col:12s} -> {n_unicos:5d} valores únicos  ({ratio:.1%} del total de filas)')

#Este código sirve para revisar cuántos valores diferentes tiene cada columna y
#que porcentaje representan del total de filas. df.columns.tolist() obtiene la
#lista de columnas, nunique() cuenta los valores únicos y ratio calcula qué porcentaje representan.


customerID   ->  7043 valores únicos  (100.0% del total de filas)
gender       ->     2 valores únicos  (0.0% del total de filas)
SeniorCitizen ->     2 valores únicos  (0.0% del total de filas)
Partner      ->     2 valores únicos  (0.0% del total de filas)
Dependents   ->     2 valores únicos  (0.0% del total de filas)
tenure       ->    73 valores únicos  (1.0% del total de filas)
PhoneService ->     2 valores únicos  (0.0% del total de filas)
MultipleLines ->     3 valores únicos  (0.0% del total de filas)
InternetService ->     3 valores únicos  (0.0% del total de filas)
OnlineSecurity ->     3 valores únicos  (0.0% del total de filas)
OnlineBackup ->     3 valores únicos  (0.0% del total de filas)
DeviceProtection ->     3 valores únicos  (0.0% del total de filas)
TechSupport  ->     3 valores únicos  (0.0% del total de filas)
StreamingTV  ->     3 valores únicos  (0.0% del total de filas)
StreamingMovies ->     3 valores únicos  (0.0% del total de filas)
Contract     ->     3 va

**3.2 — Conclusión (responde aquí en Markdown):**

¿Qué columna(s) tienen alta cardinalidad? Para la columna con mayor cardinalidad: ¿por qué nunca deberías usarla como variable predictora en un modelo, incluso si la codificaras? (relaciona tu respuesta con lo discutido en clase sobre identificadores únicos)

_Tu respuesta:_

La columna con mayor cardinalidad es  customerID, porque tiene 7043 valores únicos, es decir, un identificador diferente para cada cliente.

No debería usarse como variable predictora porque es un identificador único y no contiene información útil sobre las características del cliente. Aunque se codificara, el modelo podría aprender los identificadores en lugar de patrones reales de los datos, lo que podría causar sobreajuste y no ayudaría a predecir correctamente nuevos clientes. La segunda columna con mayor cardinalidad es TotalCharges, con 6531 valores únicos. A diferencia de customerID, no es un identificador, sino una variable que representa el total cobrado al cliente. Por eso, sí puede contener información útil para el modelo y no necesariamente debe eliminarse por tener alta cardinalidad.

**3.3 — Agrupación "Top 10 + Otros"**

En clase agrupaste `country` de Netflix Titles en sus 10 categorías más frecuentes + `'Otros'`, reduciendo su cardinalidad. Aplica la misma técnica aquí sobre la columna de mayor cardinalidad que identificaste en 3.1 (pista: `.value_counts().head(10)`, luego `.where()` + `.isin()`, igual que en el notebook de clase).

In [ ]:
# Aplica el agrupamiento Top 10 + Otros sobre la columna de mayor cardinalidad

top10 = df['TotalCharges'].value_counts().head(10)
categorias_top10 = top10.index.tolist()
df['TotalCharges_agrupado'] = df['TotalCharges'].where(
    df['TotalCharges'].isin(categorias_top10), 'Otros'
)
df['TotalCharges_agrupado'].value_counts()

#El código del ejercicio 3.3 sirve para agrupar los valores de TotalCharges:
#value_counts().head(10) encuentra los 10 valores más repetidos,
# where() mantiene esos 10 y cambia los demás por "Otros", y value_counts()
#muestra cuántos quedaron en cada grupo.



,count
TotalCharges_agrupado,
Otros,6962
20.2,11
,11
19.75,9
19.65,8
20.05,8
19.9,8
19.55,7
45.3,7


**3.4 — Conclusión (responde aquí en Markdown):**

Después de agrupar, ¿la columna resultante te parece útil para un modelo? Compara este caso con el de `country` en el notebook de clase: ¿por qué agrupar en "Top 10 + Otros" funciona bien para una variable como `country`, pero no resuelve el problema real de la columna que agrupaste aquí?

_Tu respuesta:_

La columna TotalCharges_agrupado no me parece muy útil para el modelo, porque casi todos los valores quedaron dentro de la categoría "Otros" y se perdió mucha información.

A diferencia de country, donde agrupar como "Top 10 + Otros" permite reducir muchas categorías diferentes manteniendo las más importantes, TotalCharges representa una cantidad numérica. Agruparla de esta manera hace que se pierda información sobre cuánto ha pagado realmente cada cliente. Por eso, en este caso sería mejor mantener TotalCharges como variable numérica en lugar de agruparla.

---
## Actividad 4 — Tipos de dato (30 pts)

In [ ]:
df['TotalCharges'].dtype

dtype('O')

**4.1 — Investiga (responde en Markdown):**

`TotalCharges` contiene valores numéricos (montos en dólares), pero pandas la cargó como `object`, no como `float`. Investiga por qué — revisa si hay algún valor que no se vea como un número normal.

_Tu respuesta:_

En este dataset, el problema es que TotalCharges tiene algunos valores vacíos (se puede ver en la tabla de top 10 que se tienen valores vacios). Esos valores no se pueden interpretar como números, por lo que pandas carga toda la columna como object.

In [ ]:
# 4.2 — Corrige el tipo de TotalCharges.
# Pista: pd.to_numeric() con el parámetro errors= te puede ayudar a identificar
# o manejar los valores problemáticos que encontraste en 4.1.

df["TotalCharges"]=pd.to_numeric(df["TotalCharges"], errors="coerce") #Se corrigen los valores vacios
df["TotalCharges"].dtype #se comprueba que esta bien el tipo de dato


#El código sirve para convertir TotalCharges a números, reemplazando los valores problemáticos por NaN.
#Después, dtype comprueba que la columna tenga el tipo de dato correcto

dtype('float64')

In [ ]:
# 4.3 — Convierte a category las columnas categóricas que, según lo que calculaste
# en la Actividad 3, tengan cardinalidad baja y valores fijos.
# Verifica con .dtypes que el cambio se aplicó correctamente.

#sabemos que columnas_categoricas son las que se verifico que eran tipo "object"
#asi que se aplicara el cambio ahi y se omitiran los que sabemos que tienen alta
#cadinalidad como customer ID

for columna in columnas_categoricas:
  if columna not in ['customerID', 'TotalCharges','TotalCharges_agrupado']: #Para no cambiar los identificadores de alta cardinalidad
        df[columna] = df[columna].astype('category')



print(df.dtypes)

#El código sirve para cambiar las columnas categóricas de tipo object a category,
#lo que ayuda a manejar mejor los datos categóricos. Y if evita cambiar columnas como customerID
# y TotalCharges, porque no son adecuadas para este cambio.
#El df.dtypes comprueba que las columnas se hayan cambiado correctamente

customerID                 object
gender                   category
SeniorCitizen               int64
Partner                  category
Dependents               category
tenure                      int64
PhoneService             category
MultipleLines            category
InternetService          category
OnlineSecurity           category
OnlineBackup             category
DeviceProtection         category
TechSupport              category
StreamingTV              category
StreamingMovies          category
Contract                 category
PaperlessBilling         category
PaymentMethod            category
MonthlyCharges            float64
TotalCharges              float64
Churn                    category
TotalCharges_agrupado      object
dtype: object


---
## Reflexión final (10 pts)

Con base en las 4 actividades anteriores, responde:

1. De las alertas que detectaste (formato, valores inválidos, cardinalidad, tipos), ¿cuál te pareció más fácil de decidir y cuál más difícil? ¿Por qué?
2. Si tuvieras que entregar este dataset ya "perfilado" a un compañero para que construya un modelo predictivo, ¿qué le dirías sobre `customerID` y sobre `TotalCharges`?

_Tu respuesta:_

1. La alerta que me pareció más fácil de decidir fue la de los valores inválidos, porque al revisar las columnas pude comprobar que No internet service es una categoría válida y consistente con InternetService = No. La más difícil fue la cardinalidad, porque una columna puede tener muchos valores diferentes y eso no significa necesariamente que esté mal. Hay que revisar qué representa la columna, como pasó con customerID y TotalCharges.

2. Que no use customerID como variable predictora, porque es un identificador único y no aporta información útil para predecir. En cambio, TotalCharges sí puede ser útil para el modelo porque representa el total pagado por el cliente. Sin embargo, primero debe convertirse de object a un tipo numérico (float) y revisar los valores vacíos antes de utilizarla.

---
## Rúbrica de evaluación

| Actividad | Puntos | Criterio |
|---|---|---|
| 1. Formato y valores inconsistentes | 15 | Revisó todas las columnas categóricas (no solo una); código comentado y conclusión (1.2) respaldada por lo que se observó, no solo afirmada |
| 2. Valores inválidos | 20 | Verificó la relación entre columnas antes de concluir; la conclusión (2.2) justifica con evidencia, no solo con intuición |
| 3. Alta cardinalidad | 25 | Calcula `.nunique()` para todas las columnas; aplica correctamente el agrupamiento Top 10 + Otros; la conclusión (3.4) explica por qué agrupar no resuelve el problema de un identificador único |
| 4. Tipos de dato | 30 | Identifica la causa raíz del `dtype` incorrecto (4.1); corrige `TotalCharges` sin perder información; conversión a `category` justificada por cardinalidad, no aplicada al azar |
| Reflexión final | 10 | Conecta las 4 actividades entre sí; no es una respuesta genérica o intercambiable con cualquier dataset |
| **Total** | **100** | |

**Nota sobre las conclusiones:** cada actividad tiene su propia pregunta de conclusión (1.2, 2.2, 3.4, 4.1) — esas respuestas se califican como parte de la actividad correspondiente, no solo la Reflexión final. Comenta tu código donde tomes una decisión (por ejemplo, por qué elegiste cierto umbral o cierta corrección) — el comentario también cuenta dentro del puntaje de cada actividad.